{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ddb32bb9-542e-42d4-8205-7b4a5b057e90",
   "metadata": {},
   "source": [
    "# <center>Supplementary material: calculating LD coefficients with PLINK</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13356adf",
   "metadata": {},
   "source": [
    "This notebook contains the code for the download of data from the 1000 Genomes Project and the computation of linkage disequilibrium coefficients with the software PLINK.\n",
    "\n",
    "First, we install data in the .vcf format from the 1000 Genomes Project for the three genes we are interested in (HLA-DRB1, TP53 and ABO), and save them in the \"vcf_files\" directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ac6adf6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dir.create(\"vcf_files\", showWarnings = FALSE)\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr9.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr9.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "system2(command = \"wget\", args = c(\"-c\", \"https://hgdownload.soe.ucsc.edu/gbdb/hg19/1000Genomes/phase3/ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi\"\n",
    ", \"-P\", \"./vcf_files\"))\n",
    "\n",
    "system2(command=\"wget\", args=c(\"https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel\"\n",
    ", \"-P\", \"./vcf_files\")) #this is the .panel file, useful to separate data for different populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b690fc7",
   "metadata": {},
   "source": [
    "Now we install the version 1.9 of the PLINK software."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cd5c9e42",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "url <- \"https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20231211.zip\"\n",
    "dest <- \"plink.zip\"\n",
    "\n",
    "download.file(url, dest)\n",
    "unzip(dest, exdir = \"plink\")\n",
    "\n",
    "Sys.chmod(\"plink/plink\", mode = \"0755\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cabc9649",
   "metadata": {},
   "source": [
    "We set some parameters to navigate into the 1000 Genomes Project dataset. We want data from the three genes with the starting and ending positions given by the first and last variant in the gnomAD dataset for each gene. We keep only data about African population, but we still use a vector of populations so that it is easier to add other populations if we need."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2239f960-20a1-4769-b49d-19d257145f33",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "genes <- data.frame( #data frame with the genes\n",
    "  gene  = c(\"HLA-DRB1\",\"TP53\",\"ABO\"),\n",
    "  chromosome   = c(\"6\", \"17\", \"9\"),\n",
    "  start = c(32579019, 7661882, 133233496), #start and stop taken from gnomAD dataset\n",
    "  end   = c(32589816, 7676669, 133275264),\n",
    "  stringsAsFactors = FALSE\n",
    ")\n",
    "\n",
    "populations<-c(\"AFR\") #only Africans, but we can add other populations if we want\n",
    "command<-\"./plink/plink\" #the command to use PLINK, useful later"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f1b9c4f",
   "metadata": {},
   "source": [
    "Now we write a .keep file: it will be useful later to keep only data about African population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc6f77bf-0bb7-46d2-8277-adb255b91d0d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "panel <- read.delim(\n",
    "  \"vcf_files/integrated_call_samples_v3.20130502.ALL.panel\",\n",
    "  stringsAsFactors = FALSE\n",
    ") #read the .panel file\n",
    "\n",
    "write_keep <- function(superpop, out_file) {\n",
    "  ids <- panel$sample[panel$super_pop == superpop]\n",
    "  keep <- data.frame(FID = ids, IID = ids)\n",
    "  write.table(keep, out_file,\n",
    "              quote = FALSE, row.names = FALSE,\n",
    "              col.names = FALSE, sep = \"\\t\")\n",
    "}\n",
    "\n",
    "write_keep(\"AFR\", \"AFR.keep\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8972de63",
   "metadata": {},
   "source": [
    "Using PLINK, we extract data about the three genes in African population, keeping only SNPs for positions in which we have no more than one variant, and save the results in the \"plink_out\" folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f16f023a-261f-46e6-986a-e385397ab211",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dir.create(\"plink_out\", showWarnings = FALSE) #create a new folder with the results if it does not exist\n",
    "for (i in 1:3){\n",
    "    file_path<-sprintf(\"vcf_files/ALL.chr%s.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\", genes$chromosome[i])\n",
    "    for (pop in populations){\n",
    "        output_file_name<-sprintf(\"plink_out/%s_%s\", genes$gene[i], pop) #name of the output file\n",
    "        keep_file_name<-sprintf(\"%s.keep\", pop) #file .keep for the population to keep\n",
    "        args<-c(\"--vcf\", file_path,\n",
    "                \"--keep\", keep_file_name,\n",
    "                \"--chr\", genes$chromosome[i],\n",
    "                \"--from-bp\", genes$start[i],\n",
    "                \"--to-bp\", genes$end[i],\n",
    "                \"--snps-only\", \"just-acgt\", #only SNPs\n",
    "                \"--biallelic-only\", \"strict\", #only biallelic (no more than one variant)\n",
    "                \"--make-bed\",\n",
    "                \"--out\", output_file_name)\n",
    "        system2(command, args) #system2 executes a command as in the command line\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6430a631-83cd-4242-83b9-3db32af969ad",
   "metadata": {},
   "source": [
    "We compute the LD coefficients keeping only common variants and considering only couples of positions with distance no higher than 500 kb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2e10de88-ba9e-489d-8efd-9d317ba14b6c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dir.create(\"linkage_disequilibrium\", showWarnings = FALSE)\n",
    "for (i in 1:3){\n",
    "    for (pop in populations){\n",
    "        input_file_name<-sprintf(\"plink_out/%s_%s\", genes$gene[i], pop)\n",
    "        output_file_name<-sprintf(\"linkage_disequilibrium/%s_%s_LD\", genes$gene[i], pop)\n",
    "        args<-c(\"--bfile\", input_file_name,\n",
    "                \"--maf\", \"0.01\", #only common variants (frequency higher than 1%)\n",
    "                \"--r2\", \"gz\",\n",
    "                \"--ld-window-kb\", \"500\",\n",
    "                \"--ld-window\", \"99999\",\n",
    "                \"--ld-window-r2\", \"0\", #keeps all the LD (no minimum threshold)\n",
    "                \"--out\", output_file_name\n",
    "       )\n",
    "        system2(command, args)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3485512a-ba77-472f-9caf-663033bce6f4",
   "metadata": {},
   "source": [
    "Finally, we create the .csv files associated to those data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "23d74ad7-c38d-45e8-864c-6e88d80ba114",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Warning message:\n",
      "“Missing column names filled in: 'X8' [8]”\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  CHR_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_A = \u001b[31mcol_character()\u001b[39m,\n",
      "  CHR_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_B = \u001b[31mcol_character()\u001b[39m,\n",
      "  R2 = \u001b[32mcol_double()\u001b[39m,\n",
      "  X8 = \u001b[33mcol_logical()\u001b[39m\n",
      ")\n",
      "\n",
      "Warning message:\n",
      "“Missing column names filled in: 'X8' [8]”\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  CHR_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_A = \u001b[31mcol_character()\u001b[39m,\n",
      "  CHR_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_B = \u001b[31mcol_character()\u001b[39m,\n",
      "  R2 = \u001b[32mcol_double()\u001b[39m,\n",
      "  X8 = \u001b[33mcol_logical()\u001b[39m\n",
      ")\n",
      "\n",
      "Warning message:\n",
      "“Missing column names filled in: 'X8' [8]”\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  CHR_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_A = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_A = \u001b[31mcol_character()\u001b[39m,\n",
      "  CHR_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  BP_B = \u001b[32mcol_double()\u001b[39m,\n",
      "  SNP_B = \u001b[31mcol_character()\u001b[39m,\n",
      "  R2 = \u001b[32mcol_double()\u001b[39m,\n",
      "  X8 = \u001b[33mcol_logical()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse) #for the read_table function\n",
    "\n",
    "dir.create(\"linkage_disequilibrium_csv\", showWarnings = FALSE)\n",
    "for (i in 1:3){\n",
    "    for (pop in populations){\n",
    "        input_file_name<-sprintf(\"linkage_disequilibrium/%s_%s_LD.ld.gz\", genes$gene[i], pop)\n",
    "        output_file_name<-sprintf(\"linkage_disequilibrium_csv/%s_%s_LD.csv\", genes$gene[i], pop)\n",
    "        table<-read_table(input_file_name)\n",
    "        write.csv(table, output_file_name, row.names=FALSE)\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
